test-case/volume-basic-test.sh

```diff
@@ -69,7 +69,7 @@ do
     # TODO: need to check command effect
     for kctl in "${pgalist[@]}"
     do
-        volctrl=$(echo "$kctl" | sed 's/_/ /g;')
+        volctrl=${kctl//_/ }
```

Comment:
Drop this replacement hack.

```shell
  for volctrl in "${pgalist[@]}"
```

Reply:
I tried to make changes to this variable and how it's handled, but if the replacement isn't done, amixer failes with an error.

> Yes this is more change than merely fixing shellcheck warnings but pleasing shellcheck is not just about pleasing shellcheck, sometimes it's also about making the code actually better.

I do feel like fixing the shellcheck warnings _SHOULD_ be as easy as fixing the shellcheck warnings.  The task of fixing all the shellcheck warnings has already proven monumental for what it should be, and I'd really like to be done with it. But regardless, I've been making the attempt to put your requests in place.

Script enhancements and such...  I'd be more than happy to be a reviewer and tester for any changes you'd like to add to a PR, but adding them to this PR has cost me quite a bit of time trying to get them to work.

Reply:
> I do feel like fixing the shellcheck warnings SHOULD be as easy as fixing the shellcheck warnings. 

Warnings are only a means to an end: code quality.

Reply:
`sed 's/ /_/g;'` is a whitespace hack to avoid proper quoting. Fixing the quoting without removing the hack does not really make sense.

Reply:
Oh, I see, up above we replace the spaces with underscores, so both need to be removed for this to possibly work, but the above line failed when I changed it, so I changed it back.

Reply:
This line:
`readarray -t pgalist < <(amixer -c"$sofcard" controls | awk -Fname= '/PGA/i {print $2}')`

Generates pgalist entries that have underscores:
```
declare -a pgalist=([0]="'Dmic0 Capture Switch'" [1]="'Dmic0 Capture Volume'" [2]="'Dmic1 2nd Capture Volume'" 
[3]="'EQIIR20.0 eqiir_coef_20'" [4]="'EQIIR21.0 eqiir_coef_21'" [5]="'PGA1.0 1 Master Playback Volume'" 
[6]="'PGA11.0 11 Playback Volume'" [7]="'PGA2.0 2 Master Capture Switch'" 
[8]="'PGA2.0 2 Master Capture Volume'" [9]="'PGA3.0 3 Master Playback Volume'" 
[10]="'PGA4.0 4 Master Capture Switch'" [11]="'PGA4.0 4 Master Capture Volume'" 
[12]="'PGA5.0 5 Master Playback Volume'" [13]="'PGA6.0 6 Master Capture Switch'" 
[14]="'PGA6.0 6 Master Capture Volume'" [15]="'PGA7.0 7 Playback Volume'" 
[16]="'PGA8.0 8 Playback Volume'" [17]="'PGA9.0 9 Playback Volume'")
```

Whereas this line:
`#readarray -t pgalist < <(amixer -c"$sofcard" controls | grep -i PGA | awk -Fname= '{print $2}')`
Does not.   They gather a different number of elements for the list...
```
declare -a pgalist=([0]="'PGA1.0 1 Master Playback Volume'" [1]="'PGA11.0 11 Playback Volume'" 
[2]="'PGA2.0 2 Master Capture Switch'" [3]="'PGA2.0 2 Master Capture Volume'" 
[4]="'PGA3.0 3 Master Playback Volume'" [5]="'PGA4.0 4 Master Capture Switch'" 
[6]="'PGA4.0 4 Master Capture Volume'" [7]="'PGA5.0 5 Master Playback Volume'" 
[8]="'PGA6.0 6 Master Capture Switch'" [9]="'PGA6.0 6 Master Capture Volume'" 
[10]="'PGA7.0 7 Playback Volume'" [11]="'PGA8.0 8 Playback Volume'" [12]="'PGA9.0 9 Playback Volume'")
```

Reply:
The error was unrelated to underscores, although their presense when not expected is a problem, but this is just an invalid element:
`EQIIR20.0`

Reply:
>  `awk -Fname= '/PGA/i {print $2}')`

Sorry that was wrong. I meant this: `awk -Fname= 'toupper($2) ~ /PGA/ { print $2 }'`